wsd/ClientSession.cpp

```diff
@@ -428,6 +428,90 @@ void ClientSession::onTileProcessed(TileWireId wireId)
         LOG_INF("Tileprocessed message with an unknown wire-id '" << wireId << "' from session " << getId());
 }
 
+#if !MOBILEAPP
+namespace
+{
+std::shared_ptr<http::Session> makePrepareSignatureSession(const std::shared_ptr<ClientSession> clientSession, const std::string& requestUrl)
+{
+    // Create the session and set a finished callback
+    std::shared_ptr<http::Session> httpSession = http::Session::create(requestUrl);
+    if (!httpSession)
+    {
+        LOG_WRN("PrepareSignature: failed to create HTTP session");
+        return nullptr;
+    }
+
+    http::Session::FinishedCallback finishedCallback = [clientSession](const std::shared_ptr<http::Session>& session)
+    {
+        const std::shared_ptr<const http::Response> httpResponse = session->response();
+        Poco::JSON::Object::Ptr resultArguments = new Poco::JSON::Object();
+        resultArguments->set("commandName", ".uno:PrepareSignature");
+
+        bool ok = httpResponse->statusLine().statusCode() == http::StatusCode::OK;
+        resultArguments->set("success", ok);
+
+        const std::string& responseBody = httpResponse->getBody();
+        Poco::JSON::Object::Ptr responseBodyObject = new Poco::JSON::Object();
+        if (!JsonUtil::parseJSON(responseBody, responseBodyObject))
+        {
+            LOG_WRN("PrepareSignature: failed to parse response body as JSON");
+            return;
+        }
+        resultArguments->set("result", responseBodyObject);
+
+        std::ostringstream oss;
+        resultArguments->stringify(oss);
+        std::string result = "unocommandresult: " + oss.str();
+        clientSession->sendTextFrame(result);
+    };
+    httpSession->setFinishedHandler(std::move(finishedCallback));
+    return httpSession;
+}
+}
+
+bool ClientSession::handlePrepareSignature(const StringVector& tokens)
+{
+    // Make the HTTP session: this requires an URL
+    Poco::JSON::Object::Ptr userPrivateInfoObject = new Poco::JSON::Object();
+    if (!JsonUtil::parseJSON(getUserPrivateInfo(), userPrivateInfoObject))
+    {
+        LOG_WRN("PrepareSignature: failed to parse user private info as JSON");
+        return false;
+    }
+    std::string requestUrl;
+    JsonUtil::findJSONValue(userPrivateInfoObject, "ESignatureBaseUrl", requestUrl);
+    requestUrl += "/api/signatures/prepare-files-for-signing";
+    std::shared_ptr<http::Session> httpSession = makePrepareSignatureSession(client_from_this(), requestUrl);
+    if (!httpSession)
+    {
+        return false;
+    }
+
+    // Make the request: this requires a JSON body, where we set the secret
+    std::string commandArguments = tokens.cat(' ', 2);
+    Poco::JSON::Object::Ptr commandArgumentsObject;
+    if (!JsonUtil::parseJSON(commandArguments, commandArgumentsObject))
+    {
+        LOG_WRN("PrepareSignature: failed to parse arguments as JSON");
+        return false;
+    }
+    auto requestBodyObject = commandArgumentsObject->get("body").extract<Poco::JSON::Object::Ptr>();
```

Comment:
FWIW I think (didn't verify though) that get + extract will throw if "body" doesn't exist, while getObject("body") will return null to allow an explicit test for its existence. Not sure if that unlikely case matters here, and if it does then maybe that throw is perfectly acceptable.

Reply:
This is something we control, so I would say it's OK if this throws in the unlikely case that our JS code at browser/src/control/Control.ESignature.ts:230 produces an unexpected JSON.

But you have a point here, I'll see if I can rework this to use JsonUtil::findJSONValue() instead, which won't throw, even if the key is missing. Thanks.